onmt/translate/Translator.lua

```diff
@@ -274,18 +279,32 @@ function Translator:translate(src, gold)
   if data:batchCount() > 0 then
     local batch = data:getBatch()
 
-    local pred, predFeats, predScore, attn, goldScore = self:translateBatch(batch)
+    local encStates = {}
+    local pred = {}
+    local predFeats = {}
+    local predScore = {}
+    local attn = {}
+    local goldScore = {}
+    if self.opt.print_input_embeddings then
+      encStates = self:translateBatch(batch)
+    else
+      pred, predFeats, predScore, attn, goldScore = self:translateBatch(batch)
+    end
 
     for b = 1, batch.size do
-      results[b] = {}
-
-      results[b].preds = {}
-      for n = 1, self.opt.n_best do
-        results[b].preds[n] = {}
-        results[b].preds[n].words = self:buildTargetWords(pred[b][n], src[indexMap[b]].words, attn[b][n])
-        results[b].preds[n].features = self:buildTargetFeatures(predFeats[b][n])
-        results[b].preds[n].attention = attn[b][n]
-        results[b].preds[n].score = predScore[b][n]
+      if self.opt.print_input_embeddings then
+        results[b] = encStates
```

Comment:
Each batch should have its own encoding. If `translateBatch` returns the last hidden state, it's just about doing:

```lua
results[b] = encStates[b]
```